### PR TITLE
feat(service-worker): change default SW strategy behavior

### LINKED ISSUE #34464: feat(service-worker): improve default `SwRegistrationStrategy`
**The problem**

In order to avoid negatively affecting the initial page-load time, `@angular/service-worker` will (by default) wait for the [app to stabilize][1], before registering the SW  script. This can lead to a situation where the SW will never be registered when there is a long-running task (such as an interval or recurring timeout), which can often be started by a 3rd-party dependency (without the developer realizing). Related issues: #20970, #34461

This can easily be missed by the developer and there is no other indication other than the SW is not in effect. This is particularly hard to detect, because the SW is typically not used during development and on production builds a previous SW instance might be in effect.

##
**Possible solutions**

Since 4cfba5807, it is possible to overwrite this default behavior by [providing a different `SwRegistrationStrategy`][2]. Things we could do to remedy the situation (non-exhaustive list; up for discussion):

- Change the default strategy to a custom one that wait for the app to stabilize, but registers anyway after X seconds (no matter whether stabilized or not).

- Keep the current behavior, but print a warning after X seconds that the app has not stabilized yet and the SW has not been registered. (This won't change the current behavior, but will make it easier for developers to identify the problem.)

[1]: https://angular.io/api/core/ApplicationRef#is-stable-examples
[2]: https://angular.io/api/service-worker/SwRegistrationOptions#registrationStrategy

### PR DESCRIPTION
**Pull request by sonukapoor**
Previously, there was a chance that the SW will never register
when there is a long-running task or recurring timeout. This commit
fixes this, by first trying to wait for the app to stabilize, however,
if that does not happen, then the SW will register after 30 seconds.

Closes #34464

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #34464

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by sonukapoor** [comment]
@alxhub / @gkalpak My local lint and tests are all passing - not sure why CI is complaining. Can you help with this?

**Comment by gkalpak** [comment]
Thx, @sonukapoor, for taking this on. Can you please remove the formatting changes? They make the PR hard to review and the formatter is unhappy (you are probably running the wrong commands locally if `lint` passes).

Also, how are you running the tests locally? The failures on CI seem expected.

**Comment by sonukapoor** [comment]
@gkalpak. I am running yarn lint and yarn test in the aio folder.

File: packages/service-worker/src/module.ts [hunk 1]
```diff
@@ -105,9 +105,11 @@ export function ngswAppInitializer(
           readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
           break;
         case 'registerWhenStable':
-          const appRef = injector.get<ApplicationRef>(ApplicationRef);
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
-          break;
+          // Our default is when the app is stable, or after 30 seconds
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: This is not the place to document the default behavior. This branch is about `registerWhenStable`. (The fact that we use `registerWhenStable` as the default value is done in a [different part of the code][1].)

[1]: https://github.com/angular/angular/blob/f714a54f018277c1113cbac11f5c4735e9e08da6/packages/service-worker/src/module.ts#L99

File: packages/service-worker/src/module.ts [hunk 2]
```diff
@@ -105,9 +105,11 @@ export function ngswAppInitializer(
...
+          // This ensures that the SW is still registered, in the case of a long running 
+          // or recurring timeout. 
+          readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0] || 30000)));
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Thinking about it again, I don't think we should change the behavior of `registerWhenStable` (in case it is used explicitly by someone), only the default value. It should also be possible for people to wait for stability indefinitely (if they so wish).

Therefore, I suggest the following changes:
- When not passing an argument (e.g. `'registerWhenStable'`), keep the current behavior (i.e. wait for stability indefinitely).
- When passing a timeout argument (e.g. `'registerWhenStable:10000'`), wait for stability or for the timeout to expire.
- Change the [default strategy][1] from `'registerWhenStable'` to `'registerWhenStable:30000'` (timeout up for debate).

[1]: https://github.com/angular/angular/blob/f714a54f018277c1113cbac11f5c4735e9e08da6/packages/service-worker/src/module.ts#L99

Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Hm...I just realized: Does this `delay` prevent the app from stabilizing? If that's the case (which it probably is), we should run it outside the `NgZone`. Same for other `registerWithDelay` strategy.

**Review by gkalpak — COMMENTED** [review]
Here are some more comments from a preliminary high-level review.

**Comment by gkalpak** [comment]
> @gkalpak. I am running yarn lint and yarn test in the aio folder. When I try to run similar commands within root then I get all kind of strange errors from bazel

These changes are outside the `aio/` directory, so you need to run commands at the root of the repo :grin:
Can you open a separate issue with the errors you are getting from bazel?

File: packages/service-worker/src/module.ts (hunk 2 above)
Comment (sonukapoor) [reply packages/service-worker/src/module.ts]: @gkalpak Good catch 👍

**Comment by sonukapoor** [comment]
> Can you open a separate issue with the errors you are getting from bazel?

@gkalpak, Please see here: #35891

File: packages/service-worker/src/module.ts [hunk 3]
```diff
@@ -7,8 +7,8 @@
  */
 
 import {isPlatformBrowser} from '@angular/common';
-import {APP_INITIALIZER, ApplicationRef, InjectionToken, Injector, ModuleWithProviders, NgModule, PLATFORM_ID} from '@angular/core';
-import {Observable, of } from 'rxjs';
+import {APP_INITIALIZER, ApplicationRef, InjectionToken, Injector, ModuleWithProviders, NgModule, PLATFORM_ID, NgZone} from '@angular/core';
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Please order imported symbols.

File: packages/service-worker/src/module.ts [hunk 4]
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
     if (typeof options.registrationStrategy === 'function') {
       readyToRegister$ = options.registrationStrategy();
     } else {
-      const [strategy, ...args] = (options.registrationStrategy || 'registerWhenStable').split(':');
+      // Our default is when the app is stable, or after 30 seconds
+      // This ensures that the SW is still registered, in the case of a long running 
+      // or recurring timeout. 
+      const [strategy, ...args] = (options.registrationStrategy || 'registerWhenStable:30000').split(':');
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Move this change to a separate commit.

File: packages/service-worker/src/module.ts [hunk 5]
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
...
       switch (strategy) {
         case 'registerImmediately':
           readyToRegister$ = of (null);
           break;
         case 'registerWithDelay':
-          readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          ngZone.runOutsideAngular(() => {
+            readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          });
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Have you verified that:
a. This prevents the app from stabilizng without `runOutsideangular`?
b. The app can stabilize with `runOutsideAngular` here (vs where the observable is subscribed to)?

We also need some test(s) to verify the above.

File: packages/service-worker/src/module.ts [hunk 6]
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
...
         case 'registerWhenStable':
-          const appRef = injector.get<ApplicationRef>(ApplicationRef);
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
+            ngZone.runOutsideAngular(() => {
+              if (+args[0]) {
+                readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0])));
+              } else {
+                readyToRegister$ = whenStable(injector);
+              }
+            });
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]:
```suggestion
          ngZone.runOutsideAngular(() => {
            if (+args[0]) {
              readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0])));
            } else {
              readyToRegister$ = whenStable(injector);
            }
          });
```

File: packages/service-worker/src/module.ts [hunk 7]
```diff
@@ -75,7 +75,8 @@ export const SCRIPT = new InjectionToken<string>('NGSW_REGISTER_SCRIPT');
 
 export function ngswAppInitializer(
     injector: Injector, script: string, options: SwRegistrationOptions,
-    platformId: string): Function {
+    platformId: string, 
+    ngZone: NgZone): Function {
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: In order to avoid unnecessary work, I suggest we use the `injector` to get `NgZone` only when/if needed.

File: packages/service-worker/src/module.ts [hunk 8]
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
...
-          readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          ngZone.runOutsideAngular(() => {
+            readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          });
           break;
         case 'registerWhenStable':
-          const appRef = injector.get<ApplicationRef>(ApplicationRef);
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
+            ngZone.runOutsideAngular(() => {
+              if (+args[0]) {
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: This will treat 0 the same as not providing an argument. I think this is not correct. Explicitly specifying `0` as an argument should be respected.

File: packages/service-worker/src/module.ts (hunk 6 above)
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Or rather:

```suggestion
          readyToRegister$ = (args[0] === undefined) ?
              whenStable(injector) :
              ngZone.runOutsideAngular(() => merge(whenStable(injector), of(null).pipe(delay(+args[0]))));
```

But, again, we should verify whether `runOutsideAngular()` is needed here or when subscribing to the observable.

File: packages/service-worker/src/module.ts [hunk 9]
```diff
@@ -117,13 +128,18 @@ export function ngswAppInitializer(
 
     // Don't return anything to avoid blocking the application until the SW is registered.
     // Catch and log the error if SW registration fails to avoid uncaught rejection warning.
-    readyToRegister$.pipe(take(1)).subscribe(
+    readyToRegister$!.pipe(take(1)).subscribe(
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: This should not be needed with my suggested refactoring.

File: packages/service-worker/src/module.ts [hunk 10]
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
...
+      // This ensures that the SW is still registered, in the case of a long running 
+      // or recurring timeout. 
+      const [strategy, ...args] = (options.registrationStrategy || 'registerWhenStable:30000').split(':');
       switch (strategy) {
         case 'registerImmediately':
           readyToRegister$ = of (null);
           break;
         case 'registerWithDelay':
-          readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          ngZone.runOutsideAngular(() => {
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Please move this change to a separate commit.

File: packages/service-worker/src/module.ts [hunk 11]
```diff
@@ -96,17 +97,27 @@ export function ngswAppInitializer(
...
+            readyToRegister$ = of (null).pipe(delay(+args[0] || 0));
+          });
           break;
         case 'registerWhenStable':
-          const appRef = injector.get<ApplicationRef>(ApplicationRef);
-          readyToRegister$ = appRef.isStable.pipe(filter(stable => stable));
+            ngZone.runOutsideAngular(() => {
+              if (+args[0]) {
+                readyToRegister$ = merge(whenStable(injector), of(null).pipe(delay(+args[0])));
```
Comment (gkalpak) [inline packages/service-worker/src/module.ts]: Might be worth creating a helper for these `of(null).pipe(delay(+args[0]))`.

File: packages/service-worker/test/module_spec.ts [hunk 12]
```diff
@@ -177,6 +177,21 @@ describe('ServiceWorkerModule', () => {
            expect(swRegisterSpy).toHaveBeenCalledWith('sw.js', {scope: undefined});
          }));
 
+        it('registers the SW when the app doesnt stabilizes with `registerWhenStable`, but uses the timeout', fakeAsync(() => {
```
Comment (gkalpak) [inline packages/service-worker/test/module_spec.ts]:
```suggestion
        it('registers the SW when the app does not stabilize after `<timeout>` with `registerWhenStable:<timeout>`', fakeAsync(() => {
```

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by gkalpak** [comment]
BTW, @sonukapoor, you don't need to run all the tests when you only update a specific package. For this PR for example, you could run `yarn bazel test //packages/service-worker/...` or `yarn bazel test //packages/service-worker/test`.
(Replace `bazel` with `ibazel` in the above commands to run in "watch  mode".)

File: packages/service-worker/src/module.ts (hunk 5 above)
Comment (sonukapoor) [reply packages/service-worker/src/module.ts]: I was going to test all of this using unit tests but since I can’t run bazel I am stuck on it.

**Comment by sonukapoor** [comment]
Thanks. I will do that once I get bazel working. Until then it’s very difficult to run any tests. I will make the code changes for now and test everything once the bazel issue is resolved.

File: packages/service-worker/src/module.ts (hunk 8 above)
Comment (sonukapoor) [reply packages/service-worker/src/module.ts]: Good point

File: packages/service-worker/src/module.ts (hunk 9 above)
Comment (sonukapoor) [reply packages/service-worker/src/module.ts]: I tried that, but typescript is still complaining about the error `Variable 'readyToRegister$' is used before being assigned.ts(2454)`

File: packages/service-worker/src/module.ts (hunk 6 above)
Comment (sonukapoor) [reply packages/service-worker/src/module.ts]: @gkalpak, you are right. The app will never stabilize. Here is the documentation:

https://angular.io/api/core/ApplicationRef#isstable-examples-and-caveats

> the application will never be stable if you start any kind of recurrent asynchronous task when the application starts (for example for a polling process, started with a setInterval, a setTimeout or using RxJS operators like interval);

By wrapping the code block in the `runOutsideAngular` function, we should be fine.

**Comment by sonukapoor** [comment]
@gkalpak, I think I have addressed all of your comments. However, I am not able to run any tests, because of the previously mentioned errors with bazel. I tried the single tests but get the same errors. Let me know if the code/commits are otherwise fine.

**Comment by sonukapoor** [comment]
@gkalpak, something is wrong with CI, my PWA score locally with the same arguments is at passing:

image

**Comment by sonukapoor** [comment]
@gkalpak I borrowed a mac for this PR. It seems that the tests are failing because of the timeout that we have set on the service worker. It does not complete in time with the timeout set in protractor/jasmine. It seems to be set to `10secs`. I have tried to set that but couldn't figure it out. I tried setting it directly in the `it`s third function argument or using `jasmine.DEFAULT_TIMEOUT_INTERVAL = 31000`, but nothing worked. I then simply reduced the time in the service worker to 8 secs instead of 30. Now the tests seem to pass fine - see [here](https://app.circleci.com/pipelines/github/angular/angular/11854/workflows/dd226ce1-5290-41d1-b628-d15e6084491a/jobs/658504). However, the lighthouse audits are failing now. They are passing for me on windows and mac. I am not sure why `CI` is complaining about this.

File: packages/service-worker/src/module.ts (hunk 5 above)
Comment (gkalpak) [reply packages/service-worker/src/module.ts]: It turns out that we need to `runOutsideAngular()` when subscribing to the observable (because that is where the timeouts are created). Creating the observable (but not subscribing to it) outside the zone does nothing to allow the app to stabilize :grin:

I will move the `runOutsideAngular()` call to where we subscribe.

**Comment by gkalpak** [comment]
@googlebot I consent.

**Comment by gkalpak** [comment]
Rebased on master, updated commit messages and fixed up commits. Let's see what CI thinks :smiley:
@sonukapoor, please take a look and make sure I didn't miss anything 🙏

**Comment by sonukapoor** [comment]
> Rebased on master, updated commit messages and fixed up commits. Let's see what CI thinks 😃
> @sonukapoor, please take a look and make sure I didn't miss anything 🙏

Looks good to me. Thanks for the updates - very much appreciated.

**Review by gkalpak — APPROVED** [review]
Thx, @sonukapoor ✨
